crates/core/src/operations/add_column.rs

```diff
@@ -0,0 +1,125 @@
+//! Add a new column to a table
+
+use futures::future::BoxFuture;
+use itertools::Itertools;
+use maplit::hashset;
+
+use super::cast::merge_struct;
+use super::transaction::{CommitBuilder, CommitProperties, PROTOCOL};
+
+use crate::kernel::{Protocol, ReaderFeatures, StructField, WriterFeatures};
+use crate::logstore::LogStoreRef;
+use crate::operations::set_tbl_properties::convert_properties_to_features;
+use crate::protocol::DeltaOperation;
+use crate::table::state::DeltaTableState;
+use crate::{DeltaResult, DeltaTable, DeltaTableError};
+
+/// Add new columns and/or nested fields to a table
+pub struct AddColumnBuilder {
+    /// A snapshot of the table's state
+    snapshot: DeltaTableState,
+    /// Fields to add/merge into schema
+    fields: Option<Vec<StructField>>,
+    /// Delta object store for handling data files
+    log_store: LogStoreRef,
+    /// Additional information to add to the commit
+    commit_properties: CommitProperties,
+}
+
+impl super::Operation<()> for AddColumnBuilder {}
+
+impl AddColumnBuilder {
+    /// Create a new builder
+    pub fn new(log_store: LogStoreRef, snapshot: DeltaTableState) -> Self {
+        Self {
+            snapshot,
+            log_store,
+            fields: None,
+            commit_properties: CommitProperties::default(),
+        }
+    }
+
+    /// Specify the fields to be added
+    pub fn with_fields(mut self, fields: Vec<StructField>) -> Self {
```

Comment:
I recommend making fields `an impl IntoIter` sort of thing for more flexibility by the caller